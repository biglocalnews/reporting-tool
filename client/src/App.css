@import "~antd/dist/antd.css";

.site-layout-background {
  background: #fff;
  padding: 40px;
  margin-top: 24px;
  min-height: 280px;
}
