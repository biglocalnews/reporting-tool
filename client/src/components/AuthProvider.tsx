import React, { useContext } from "react";
import { Auth } from "../services/auth";

/**
 * Context containing auth object.
 */
const AuthContext = React.createContext<Auth | null>(null);

/**
 * Component to hook into suspense and provide the given auth object.
 */
export const AuthProvider = ({ auth, children }: any) => {
  if (auth.initState !== "ready") {
    throw auth.init();
  }

  return <AuthContext.Provider value={auth}>{children}</AuthContext.Provider>;
};

/**
 * React hook to get the auth object from context.
 */
export const useAuth = () => useContext(AuthContext)!;
