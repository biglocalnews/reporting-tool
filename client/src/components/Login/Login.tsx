import React, { useState } from "react";
import { useTranslation } from "react-i18next";
import { Button, Form, Card, Input, Typography } from "antd";
import { RouteComponentProps } from "react-router-dom";
import { Auth } from "../../services/auth";
import "./Login.css";

const { Text } = Typography;

/**
 * Data passed by the user into a login attempt.
 */
type LoginRequest = {
  email: string;
  password: string;
};

/**
 * Props expected by a login component.
 *
 * Mostly these come from react-router, but the auth component is our custom
 * service that manages authentication.
 */
export type LoginProps = { auth: Auth } & RouteComponentProps;

/**
 * Login UI form.
 */
export const Login = (props: LoginProps) => {
  const { t, i18n } = useTranslation();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<Error | null>(null);

  const onFinish = async ({ email, password }: LoginRequest) => {
    setError(null);

    const error = await props.auth.login(email, password);
    if (error) {
      setError(new Error(error));
    } else {
      const state = props.location.state;
      const redirect = (state && (state as Record<string, any>).from) || {
        pathname: "/",
      };
      props.history.push(redirect);
    }
  };

  const onFinishFailed = (...args: any[]) => {
    console.log("finish failed", args);
  };

  return (
    <div className="login">
      <Card className="login-form">
        <Form
          name="login"
          labelCol={{ span: 8 }}
          wrapperCol={{ span: 16 }}
          onFinish={onFinish}
          onFinishFailed={onFinishFailed}
        >
          <Form.Item wrapperCol={{ offset: 8, span: 16 }}>
            <Text>{t("loginPrompt")}</Text>
          </Form.Item>
          <Form.Item
            label={t("enterEmailLabel")}
            name="email"
            rules={[
              {
                required: true,
                message: t("enterEmailError"),
              },
            ]}
          >
            <Input
              type="email"
              aria-label="e-mail"
              aria-required={true}
              required
            />
          </Form.Item>
          <Form.Item
            label={t("enterPasswordLabel")}
            name="password"
            rules={[
              {
                required: true,
                message: t("enterPasswordError"),
              },
            ]}
          >
            <Input.Password
              aria-label="password"
              aria-required={true}
              required
            />
          </Form.Item>
          {error && (
            <Form.Item wrapperCol={{ offset: 8, span: 16 }}>
              <Text>Error: {error.message}</Text>
            </Form.Item>
          )}
          <Form.Item wrapperCol={{ offset: 8, span: 16 }}>
            <Button type="primary" htmlType="submit">
              {t("loginAction")}
            </Button>
          </Form.Item>
        </Form>
      </Card>
    </div>
  );
};
