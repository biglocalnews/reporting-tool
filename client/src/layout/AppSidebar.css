.ant-layout-sider {
  background: #fff !important;
}

.ant-menu-dark .ant-menu-inline.ant-menu-sub {
  background: unset;
}

.ds-container > ul.ant-menu-sub {
  max-height: 60vh;
  overflow: auto;
  scrollbar-width: none;
}

/* width */
::-webkit-scrollbar {
  width: 8px;
}

/* Track */
::-webkit-scrollbar-track {
  box-shadow: inset 0 0 5px grey;
  border-radius: 8px;
}

/* Handle */
::-webkit-scrollbar-thumb {
  background: #2f5e8a;
  border-radius: 8px;
}
