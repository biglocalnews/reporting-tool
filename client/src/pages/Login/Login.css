.login {
  width: 100%;
  height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
}

.login-form {
  max-width: 600px;
  flex-basis: 600px;
}
