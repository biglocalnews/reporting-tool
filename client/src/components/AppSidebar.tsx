import React from "react";
import "../App.css";
import { Layout, Menu } from "antd";
import { TeamOutlined, BarChartOutlined } from "@ant-design/icons";
import { useTranslation } from "react-i18next";
import i18next from "../i18n/i18next";
import { Link } from "react-router-dom";

const { SubMenu } = Menu;
const { Sider } = Layout;

// TODO: add gql query for retrieving datasets a user has access to
const programs: any = [
  {
    key: "1",
    team: "BBC News",
    datasets: [
      {
        key: "1",
        id: 1,
        title: "Instagram",
      },
      {
        key: "2",
        id: 2,
        title: "12pm-4pm",
      },
      {
        key: "3",
        id: 3,
        title: "Breakfast Hour",
      },
    ],
  },
];

const AppSidebar = () => {
  const { t, i18n } = useTranslation();

  return (
    <Sider width="auto" className="sidebar" breakpoint="md">
      <Menu
        mode="inline"
        theme="light"
        defaultOpenKeys={["teams", "stats"]}
        style={{ height: "100%", borderRight: 0 }}
      >
        <SubMenu
          key="teams"
          title={t("teamsSideBarTitle")}
          icon={<TeamOutlined />}
        >
          {programs.map(
            (program: { key: string; team: string; datasets: any[] }) => {
              return (
                <Menu.ItemGroup key={program.key} title={program.team}>
                  {program.datasets.map((dataset) => (
                    <Menu.Item key={dataset.key}>
                      <Link
                        to={{
                          pathname: `/dataset/${dataset.id}/details`,
                        }}
                      >
                        {dataset.title}
                      </Link>
                    </Menu.Item>
                  ))}
                </Menu.ItemGroup>
              );
            }
          )}
        </SubMenu>
        <SubMenu key="stats" title="My Stats" icon={<BarChartOutlined />}>
          <div style={{ padding: "20px", background: "#fff" }}>Chart here</div>
        </SubMenu>
      </Menu>
    </Sider>
  );
};

export { AppSidebar };
