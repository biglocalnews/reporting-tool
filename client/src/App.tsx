import { Divider, Empty, Layout } from "antd";
import { Suspense, useEffect, useState } from "react";
import { BrowserRouter, Navigate, Outlet, Route, Routes, useLocation } from "react-router-dom";
import "./App.css";
import { ErrorBoundary } from "./components/Error/ErrorBoundary";
import { Loading } from "./components/Loading/Loading";
import AppHeader from "./layout/AppHeader";
import { AppSidebar } from "./layout/AppSidebar";
import { EditProgram } from "./pages/Admin/EditProgram";
import { EditTags } from "./pages/Admin/EditTags";
import { EditTeam } from "./pages/Admin/EditTeam";
import { EditUser } from "./pages/Admin/EditUser";
import { ProgramList } from "./pages/Admin/ProgramList";
import { TeamList } from "./pages/Admin/TeamList";
import { UserList } from "./pages/Admin/UserList";
import { DataEntry } from "./pages/DataEntry/DataEntry";
import { DatasetDetails } from "./pages/DatasetDetails/DatasetDetails";
import { Login } from "./pages/Login/Login";
import { useAuth } from "./components/AuthProvider";
import React from "react";
import { useTranslation } from "react-i18next";
import { Home } from "./pages/Home/Home";
import { Datasets } from "./pages/Datasets/Datasets";
import { Reports } from "./pages/Reports/Reports";
import { AdminReports } from "./pages/Admin/AdminReports";
const { Footer, Content } = Layout;




function Redirecter(props: { from: string }) {
  useEffect(() => {
    console.log(props.from);
    window.location.href = `/api/bbc-login`;
  });
  return <React.Fragment />;
}

const NotFound = () => {
  const { t } = useTranslation();
  return <div style={{ display: 'inline-flex', justifyContent: 'center', alignItems: 'center', height: "100%" }}>
    <Empty
      description={
        <span>{t("pageNotFound")}</span>
      }
    >
    </Empty>
  </div>
}

/**
 * Top-level app layout.
 */
function App() {
  const [sidebarCollapsed, setSidebarCollapse] = useState(false);
  const footerHeight = "48px";
  /**
 * Layout container for an authenticated user.
 */
  const ProtectedAppContainer: React.FC = ({ children }) =>
    <Layout hasSider>
      <AppSidebar setCollapseState={setSidebarCollapse} collapsed={sidebarCollapsed} />
      <Layout
        style={{
          backgroundColor: "#fff",
          marginLeft: sidebarCollapsed ? 80 : 300,
          overflowY: "auto",
          overflowX: "auto",
          height: `calc(100vh - ${footerHeight})`,
          width: `calc(100vw - ${sidebarCollapsed ? 80 : 300}px`,
        }}
      >
        <AppHeader />
        <Divider style={{ margin: 0, borderTop: "15px solid #f0f2f5" }} />
        <Content
          style={{
            padding: "20px",
            height: "auto",
            width: "100%"
          }}
        >
          {children}
        </Content>
        <Footer
          style={{
            padding: "14px 50px",
            height: footerHeight,
            textAlign: 'center',
            position: 'fixed',
            zIndex: 1,
            bottom: 0,
            boxShadow: "0.5em 0.5em 0.5em rgba(0,0,50,0.6)",
            width: `calc(100vw - ${sidebarCollapsed ? 80 : 300}px`
          }}
        >
          BBC Northern Ireland Products Team et Stanford Fecit
        </Footer>
      </Layout>
    </Layout>

  // Additional routes that only authed admins can visit.
  // When a user is logged in but lacks permission, they will see an error
  // telling them that they can't view the requested site.
  //
  // If a user is *not* logged in, they will be redirected to login screen.
  const WrappedPrivateAdmin = () => {
    const location = useLocation();
    const auth = useAuth();
    const { t } = useTranslation();
    if (!auth.isLoggedIn()) {
      if (process.env.NODE_ENV !== "production") {
        return <Navigate to="/login" state={{ from: location }} />
      }
      return <Redirecter from={location.pathname} />
    }
    return <ProtectedAppContainer>
      <ErrorBoundary>
        {auth.isAdmin() ? (
          <Outlet />
        ) : (
          <div>{t("notAuthorized")}</div>
        )}
      </ErrorBoundary>
    </ProtectedAppContainer>
  }



  // Routes that a normal authed user can visit.
  // If the user is not logged in, they will be redirected to the login screen.
  const WrappedPrivate = () => {
    const location = useLocation();
    const auth = useAuth();
    if (!auth.isLoggedIn()) {
      if (process.env.NODE_ENV !== "production") {
        return <Navigate to="/login" state={{ from: location }} />
      }
      return <Redirecter from={location.pathname} />
    }
    return <ProtectedAppContainer>
      <ErrorBoundary>
        <Outlet />
      </ErrorBoundary>
    </ProtectedAppContainer>

  }


  return (
    <Suspense fallback={<Loading />}>
      <BrowserRouter>
        <Routes>
          <Route path="/login" element={<Login />} />
          <Route path="/" element={<WrappedPrivate />}>
            <Route index element={<Home />} />
            <Route path="dataset/:datasetId/details" element={<DatasetDetails />} />
            <Route path="dataset/:datasetId/entry" element={<DataEntry />} />
            <Route path="dataset/:datasetId/entry/edit/:recordId" element={<DataEntry />} />
            <Route path="reports" element={<Reports />} />
          </Route>
          <Route path="/admin/" element={<WrappedPrivateAdmin />}>
            <Route path="users" element={<UserList />} />
            <Route path="users/:userId" element={<EditUser />} />
            <Route path="teams" element={<TeamList />} />
            <Route path="teams/:teamId" element={<EditTeam />} />
            <Route path="programs" element={<ProgramList />} />
            <Route path="programs/:programId" element={<EditProgram />} />
            <Route path="tags" element={<EditTags />} />
            <Route path="datasets" element={<Datasets />} />
            <Route path="reports" element={<AdminReports />} />
          </Route>
          <Route path="*" element={<NotFound />} />
        </Routes>
      </BrowserRouter>
    </Suspense>
  );
}

export default App;
