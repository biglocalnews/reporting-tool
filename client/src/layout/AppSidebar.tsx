import {
  BarChartOutlined,
  DatabaseOutlined,
  SettingOutlined,
  TableOutlined,
  TeamOutlined,
  UserSwitchOutlined,
} from "@ant-design/icons";
import { useQuery } from "@apollo/client";
import { Layout, Menu } from "antd";
import { useTranslation } from "react-i18next";
import { Link } from "react-router-dom";
import { useAuth } from "../components/AuthProvider";
import { GetUser, GetUserVariables } from "../graphql/__generated__/getUser";
import { GET_USER } from "../graphql/__queries__/GetUser.gql";
import "./AppSidebar.css";
const { SubMenu } = Menu;
const { Sider } = Layout;


/**
 * Sidebar content for all users.
 */
export const AppAdminSidebarMenu = () => {
  const { t } = useTranslation();
  return (
    <Menu mode="inline" role="menubar">
      <Menu.Item key="alldata" icon={<DatabaseOutlined />} role="menuitem">
        <Link to="/">{t("admin.sidebar.viewAll")}</Link>
      </Menu.Item>
      <Menu.ItemGroup key="admin" title={t("admin.sidebar.controlsHeader")}>
        <Menu.Item key="users" icon={<UserSwitchOutlined />} role="menuitem">
          <Link to="/admin/users">{t("admin.sidebar.manageUsers")}</Link>
        </Menu.Item>
        <Menu.Item key="teams" icon={<TeamOutlined />} role="menuitem">
          <Link to="/admin/teams">{t("admin.sidebar.manageTeams")}</Link>
        </Menu.Item>
        <Menu.Item key="programs" icon={<TableOutlined />} role="menuitem">
          <Link to="/admin/programs">{t("admin.sidebar.managePrograms")}</Link>
        </Menu.Item>
        <Menu.Item key="tags" icon={<TableOutlined />} role="menuitem">
          <Link to="/admin/tags">{t("admin.sidebar.manageTags")}</Link>
        </Menu.Item>
      </Menu.ItemGroup>
    </Menu>
  );
};

export const AppSidebarMenu = () => {
  const { t } = useTranslation();
  const userId = useAuth().getUserId();
  const auth = useAuth();
  const { data, loading } = useQuery<GetUser, GetUserVariables>(GET_USER, {
    variables: { id: userId },
  });

  const sidebarTeams = data?.user?.teams.flatMap((team) => {
    return {
      key: team.id,
      name: team.name,
      programmes:
        team.programs.map((program) => {
          return {
            key: program.id,
            team: program.name,
            datasets: program.datasets.map((dataset) => {
              return {
                id: dataset.id,
                title: dataset.name,
              };
            }),
          };
        })
    };
  }


  );

  return (
    <Menu
      mode="inline"
      theme="dark"
      /*defaultOpenKeys={["admin"]}*/
      style={{ height: "100%", borderRight: 0, paddingTop: "10px" }}
    >
      {auth.isAdmin() ? (
        <Menu.Item key="alldata" icon={<DatabaseOutlined />} role="menuitem">
          <Link to="/">{t("admin.sidebar.viewAll")}</Link>
        </Menu.Item>
      ) : (
        ""
      )}


      <SubMenu
        key="my-teams"
        title="My Teams"
        icon={<TeamOutlined />}
      >
        {loading ? (
          <h1>Loading...</h1>
        ) : (
          sidebarTeams?.map((item) => {
            return (
              <SubMenu
                className="ds-container"
                key={item.key}
                title={item.name}
              >
                {item.programmes
                  .sort((a, b) => a.team.localeCompare(b.team))
                  .map((item) => (
                    <SubMenu
                      className="ds-container"
                      key={item.key}
                      title={item.team}
                    >
                      {item.datasets.map((dataset) => (
                        <Menu.Item key={dataset.id}>
                          <Link
                            to={{
                              pathname: `/dataset/${dataset.id}/details`,
                            }}
                          >
                            {dataset.title}
                          </Link>
                        </Menu.Item>
                      ))

                      }

                    </SubMenu>
                  ))}
              </SubMenu>
            );
          }
          )
        )}
      </SubMenu>

      {/*<SubMenu key="stats" title="My Stats" icon={<BarChartOutlined />}>
        <div style={{ padding: "20px", background: "#fff" }}></div>
          </SubMenu>*/}

      {auth.isAdmin() ? (
        <SubMenu key="stats" title="Reports" icon={<BarChartOutlined />}>
          <Menu.Item key="reports" role="menuitem">
            <Link to="/admin/reports">{t("admin.sidebar.reports")}</Link>
          </Menu.Item>
        </SubMenu>
      ) : (
        ""
      )}

      {auth.isAdmin() ? (
        <SubMenu key="admin" title="Admin" icon={<SettingOutlined />}>
          <Menu.Item key="users" role="menuitem">
            <Link to="/admin/users">{t("admin.sidebar.manageUsers")}</Link>
          </Menu.Item>
          <Menu.Item key="teams" role="menuitem">
            <Link to="/admin/teams">{t("admin.sidebar.manageTeams")}</Link>
          </Menu.Item>
          <Menu.Item key="programs" role="menuitem">
            <Link to="/admin/programs">
              {t("admin.sidebar.managePrograms")}
            </Link>
          </Menu.Item>
          <Menu.Item key="tags" role="menuitem">
            <Link to="/admin/tags">
              {t("admin.sidebar.manageTags")}
            </Link>
          </Menu.Item>
        </SubMenu>
      ) : (
        ""
      )}
    </Menu>
  );
};

/**
 * App sidebar content: info and navigation links
 */
const AppSidebar = (): JSX.Element => {
  return (
    <Sider
      width={300}
      className="sidebar"
      breakpoint="md"
      theme="dark"
      collapsible
    >
      <AppSidebarMenu />
      {/*auth.isAdmin() ? <AppAdminSidebarMenu /> : <AppNormalUserSidebarMenu />*/}
    </Sider>
  );
};

export { AppSidebar };
