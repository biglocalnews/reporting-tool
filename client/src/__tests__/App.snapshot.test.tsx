import React from "react";
import { MemoryRouter } from "react-router-dom";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import App, { ProtectedAppContainer } from "../App";
import {
  mockUserLoggedIn,
  mockUserNotLoggedIn,
} from "../graphql/__mocks__/auth";
import { AuthProvider } from "../components/AuthProvider";
import { Login } from "../components/Login/Login";
import { AppSidebar, AppAdminSidebarMenu } from "../layout/AppSidebar";

it("renders App correctly", async () => {
  const { auth, mock } = mockUserLoggedIn();
  await auth.init();

  const tree = shallow(<App />, {
    wrappingComponent: AuthProvider,
    wrappingComponentProps: { auth },
  });

  expect(toJson(tree)).toMatchSnapshot();
});

it("renders Login when not authed", async () => {
  const { auth, mock } = mockUserNotLoggedIn();
  await auth.init();

  const tree = mount(<App />, {
    wrappingComponent: AuthProvider,
    wrappingComponentProps: { auth },
  });

  const login = tree.find(Login);
  expect(login).toBeTruthy();
});

test("renders ProtectedAppContainer correctly", async () => {
  const tree = shallow(
    <ProtectedAppContainer>
      <div />
    </ProtectedAppContainer>
  );
  expect(toJson(tree)).toMatchSnapshot();
});

it("renders admin stuff when authed as admin", async () => {
  const { auth, mock } = mockUserLoggedIn({
    roles: [{ name: "admin", description: "" }],
  });
  await auth.init();

  const tree = shallow(<App />, {
    wrappingComponent: AuthProvider,
    wrappingComponentProps: { auth },
  });

  expect(toJson(tree)).toMatchSnapshot();
});

it("renders custom sidebar for admin", async () => {
  const { auth, mock } = mockUserLoggedIn({
    roles: [{ name: "admin", description: "" }],
  });
  await auth.init();

  const tree = mount(
    <MemoryRouter>
      <AuthProvider auth={auth}>
        <AppSidebar />
      </AuthProvider>
    </MemoryRouter>
  );

  const adminSidebar = tree.find(AppAdminSidebarMenu);
  expect(adminSidebar).toBeTruthy();
});
