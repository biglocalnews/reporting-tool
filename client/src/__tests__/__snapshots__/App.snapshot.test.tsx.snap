// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Suspense
  fallback={
    <h1>
      Loading profile...
    </h1>
  }
>
  <Layout
    style={
      Object {
        "height": "100vh",
      }
    }
  >
    <AppHeader />
    <Layout>
      <AppSidebar />
      <Layout
        style={
          Object {
            "padding": "0 24px 24px",
          }
        }
      >
        <Content
          className="site-layout-background"
          style={
            Object {
              "marginTop": 24,
              "minHeight": 280,
              "padding": 24,
            }
          }
        />
        <Footer
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Footer content
        </Footer>
      </Layout>
    </Layout>
  </Layout>
</Suspense>
`;
